### ISSUE TITLE
Should we split the charging functionality into a dedicated hook

### ISSUE THREAD

**Issue by MayeulC** [postmarketOS/charging-sdl#20]
I asked about thin in #postmarketos, and @PureTryOut wasn't too keen on this.

However, I would like to receive additional feedback.

Currently, everything that's not related to the actual stuff displayed on the screen is handled here:
https://github.com/postmarketOS/pmbootstrap/blob/2824a525746c0a59b5216c73413ac0179faa0e97/aports/main/postmarketos-mkinitfs/init.sh.in#L35-L38
and there:
https://github.com/postmarketOS/pmbootstrap/blob/2824a525746c0a59b5216c73413ac0179faa0e97/aports/main/postmarketos-mkinitfs/init_functions.sh#L288-L309

As some logic needs to be added there, as per most issues in this repository, I expect the corresponding function to grow in size. Shouldn't the whole charging functionality be split into a dedicated kernel hook? This is mainly for consistency, and as a way to have the feature self-contained.

As a bonus, charging-sdl would be optional (though included by default), if unwanted (it *might* crash on some devices, or someone might just dislike it, who knows).

The downside is that we will probably need to duplicate the following calls:
https://github.com/postmarketOS/pmbootstrap/blob/2824a525746c0a59b5216c73413ac0179faa0e97/aports/main/postmarketos-mkinitfs/init.sh.in#L27-L36

An alternative to duplicating those would be to have secondary hooks that are used only in the second initramfs after it has loaded, which could be useful for other functionality as well. Although, I must ask if not starting USB network wouldn't help #16 ?

I am interested in addressing the other issues, but would rather get this one out of the way before starting.

**Comment by opendata26** [comment]
it could be something like the init script just checks cmdline and then launches charging sdl, which can handle backlight and button detection itself

**Comment by opendata26** [comment]
Yeah... I do think it would be nice to have input handling functionality though as it would kinda make it more portable

**Comment by MayeulC** [comment]
Well, I am actually not sure it would make it more portable... I haven't looked at triggerhappy yet, but it sounds like its functionality would be quite hard to replicate properly.

So, I went ahead and started to address one of the issues with C code (I'm a bit rusty -- by all meanings -- so please bear with my code) and I have noticed the following issues with doing everything in C :
* Trivial file manipulation is more complicated (which is a shame for system operations... have a look at my initial patch for writing into two different files: https://github.com/MayeulC/charging-sdl/commit/22b49563bd62f977a62f09bc95b2c9d2dcbbf196 )
* It might make testing more complicated, as we now probably have to include additional command line switches to disable backlight deactivation (the program contains more than a UI, so it is).
* The barrier to entry for new contributors is probably higher (I've seen that many people don't know C very well, though that would probably be enough for minor fixes).

However, it also has some other advantages, like the fact that it is self-contained. I would like someone else to weight on this.

**Comment by opendata26** [comment]
SDL should provide stuff for backlight and input events so it should be pretty simple

**Comment by ollieparanoid** [comment]
I have moved the "perform everything in the C code" discussion to #21.

Back to @MayeulC's original question, whether to move `start_charging_mode()` into its own hook file.

> As some logic needs to be added there, as per most issues in this repository, I expect the corresponding function to grow in size. Shouldn't the whole charging functionality be split into a dedicated kernel hook? This is mainly for consistency, and as a way to have the feature self-contained.

Yes, I think this is a great idea :+1:

> An alternative to duplicating those would be to have secondary hooks that are used only in the second initramfs after it has loaded

That sounds nice! I've written down some untested dummy code, and noticed that we could even add another set of hooks right before we switch the root, so we can get rid of the `killall telnetd msm-fb-refresher` code that does not make sense when the debug-shell or msm-fb-refresher hook are not installed:

`init_functions.sh`:
```shell
...
# $1: hook folder suffix
run_hooks() {
	local dir = "/etc/postmarketos-mkinitfs/hooks$1"
	for hook in "$dir"/*.sh; do
		[ -e "$hook" ] || continue
		echo "Run hook $hook"
		sh "$hook"
	done
}
...
```

`init.sh.in`:
```shell
...
# Fix for framebuffer drivers, which do not have a default mode
set_framebuffer_mode

# Run hooks
run_hooks
mount_boot_partition
extract_initramfs_extra @INITRAMFS_EXTRA@
run_hooks "-extra"

# Setup USB network
setup_usb_network
start_udhcpd

# Setup root partition
wait_root_partition
resize_root_partition
unlock_root_partition
resize_root_filesystem
mount_root_partition

# Switch root
run_hooks "-switch-root"
killall mdev
umount /boot
umount /proc
...
```

@drebrez: you've written quite some initramfs code, what's your opinion on this?
@PureTryOut: can you share the reasons why you were against this change?

(@MayeulC: I remember that you've asked about how to reference source code in GitHub, so it shows the code directly. This only works in the same repository for some reason, which means you can reference pmbootstrap code in the pmbootstrap issues only.)

**Comment by drebrez** [comment]
@ollieparanoid I think it's a good idea to reduce the initramfs code and extend it's functionality with additional hooks. 
The draft code you wrote looks pretty good 👍 
Should we need to consider also to have a list of hooks that have to be installed by default?
I'm thinking about if it's better to have the charging-sdl hook always installed but with the possibility to remove it or leave the choice to the user, maybe even with a config option in `pmbootstrap init`?

**Comment by ollieparanoid** [comment]
Thanks for taking a look at it!

I would always install charging-sdl for now, until we have an use case where someone would actually not want to have it. Then I'd go with the list of default hooks (I guess you meant in `pmb/config/__init__.py`?) and ask the user in `pmbootstrap init`. :+1:

**Comment by MayeulC** [comment]
Thanks @ollieparanoid, I like how you approached the idea (also thanks for the tip about including code, it might have made more sense to open this issue in pmbootstrap).

I would install it by default as well, maybe installing it as a dependency of the device package. Does Alpine have meta-packages? We could make `postmarketos-hooks-defaults` a dependency of most devices. This causes some problems when one actually wants to disable a hook, but maybe we shouldn't worry too much about this until there's an actual use case, as @ollieparanoid pointed out.

I can't remember right now how exactly hooks are handled in pmbootsrap, so it may be that the previous paragraph doesn't apply. My point is that I don't really think this is something we should ask in `pmbootstrap init`, as:
* it might confuse first-time users
* the use case for customization is likely device-specific
* if the user wants to customize it, `hook_add` can always be used
* I would rather not let the list of questions asked by `pmbootstrap init` grow to unmanageable size

We could however display some information: hooks installed, and a link to a wiki page describing what hooks are, how they work and which ones exist.

**Comment by z3ntu** [comment]
Maybe an "expert mode" in the init where you can pick and choose the hooks and in the non-expert mode it chooses sane defaults? Would be a good option for all imo.

**Comment by ollieparanoid** [comment]
> Does Alpine have meta-packages?

Yes, e.g. [xfce4](https://pkgs.alpinelinux.org/package/edge/community/x86_64/xfce4).

> We could however display some information: hooks installed, and a link to a wiki page describing what hooks are, how they work and which ones exist.

Good idea.

> Maybe an "expert mode" in the init where you can pick and choose the hooks and in the non-expert mode it chooses sane defaults? Would be a good option for all imo.

Sounds good. We have the "build options" (see below) which already collapses a lot of questions, we could rename that to "expert options" and add the initramfs hooks question there if needed.

```
Build options: Parallel jobs: 3, ccache per arch: 5G
Change them? (y/n) [n]:
```

**Comment by drebrez** [comment]
@ollieparanoid what you think about creating a dedicated issue in pmbootstrap repository to track the split of hooks like you suggested?
